File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
class Solution {
private:
    int solve(int &maxSum, TreeNode* root){
        if(root == nullptr) return 0;

        int l = max(0,solve(maxSum, root->left));
        int r = max(0,solve(maxSum, root->right));

        maxSum = max(maxSum, l + r + root->val);
        return max(l , r) + root->val;
    }
public:
    int maxPathSum(TreeNode* root) {
        if(!root) return 0;
        int maxSum = INT_MIN;
        solve(maxSum, root);
        return maxSum;
    }
};	

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
class Solution:
    def maxPathSum(self, root: TreeNode | None) -> int:
        maxi = float('-inf')

        def dfs(root):
            nonlocal maxi

            if root is None:
                return 0

            l = max(0, dfs(root.left))
            r = max(0, dfs(root.right))

            # Path that uses both sides through this node
            maxi = max(maxi, root.val + l + r)

            # Path that can be extended by the parent
            return root.val + max(l, r)

        dfs(root)
        return maxi